test/test_torch.py

```diff
@@ -12828,6 +12855,179 @@ def test_function_unwrap_message(self):
+            ("round", doubles, True, True, 'cpu'),
+            ("round", doubles, False, True, 'cuda'),
+            ("rsqrt", positives, True, True, 'cpu'),
+            ("rsqrt", positives, False, True, 'cuda'),
+            ("sin", doubles, True, True, 'cpu'),
+            ("sin", doubles, False, True, 'cuda'),
+            ("sigmoid", doubles, True, True, 'cpu'),
+            ("sigmoid", doubles, False, False, 'cuda'),
+            ("sqrt", doubles, True, True, 'cpu'),
+            ("sqrt", doubles, False, True, 'cuda'),
+            ("tan", doubles, True, True, 'cpu'),
+            ("tan", doubles, False, True, 'cuda'),
+            ("tanh", doubles, True, True, 'cpu'),
+            ("tanh", doubles, False, True, 'cuda'),
+            ("trunc", doubles, True, True, 'cpu'),
+            ("trunc", doubles, False, True, 'cuda')
+        ]
+
+        def _test(fn, inputs, has_input_output_mem_overlap_check,
+                  has_internal_mem_overlap_check, dev):
+            if dev != device:
+                return
+            outplace_fn = getattr(torch, fn)
+            inplace_fn = getattr(torch.Tensor, fn + '_')
+            if has_input_output_mem_overlap_check:
+                self.unary_check_input_output_mem_overlap(self, inputs, sz,
+                                                          outplace_fn)
+            else:
+                with self.assertRaises(AssertionError):
+                    self.unary_check_input_output_mem_overlap(self, inputs, sz,
+                                                              outplace_fn)
+            if has_internal_mem_overlap_check:
+                self.unary_check_internal_mem_overlap(self, inplace_fn,
+                                                      device=dev)
+            else:
+                with self.assertRaises(AssertionError):
+                    self.unary_check_internal_mem_overlap(self, inplace_fn,
+                                                          device=dev)
+
+        for test_case in unary_mem_overlap_cases:
+            _test(*test_case)
+
+    @torchtest.for_all_device_types()
+    def test_binary_op_mem_overlap(self, device):
+        ops = [
+            (torch.add, torch.Tensor.add_),
+            (torch.mul, torch.Tensor.mul_),
+            (torch.sub, torch.Tensor.sub_),
+            (torch.div, torch.Tensor.div_)
+        ]
+        for (out_op, inplace_op) in ops:
+            # internal overlap check for inplace_op
+            self.binary_check_internal_mem_overlap(self, inplace_op,
+                                                   device=device)
+            # input/output overlap check for out op
+            self.binary_check_input_output_mem_overlap(self, out_op, device)
+
+    @torchtest.for_all_device_types()
+    def test_addXXX_mem_overlap(self, device):
+        ops = [
+            (torch.addcmul, torch.Tensor.addcmul_),
+            (torch.addcdiv, torch.Tensor.addcdiv_)
+        ]
+        for (out_op, inplace_op) in ops:
```

Comment:
A lot of this test is repetitive with the binary_op test. Here's an idea to save more space while testing ternary ops. At the end of the day, here's what we want to be testing:
- the written-to tensor doesn't have internal memory overlap (for out-of-place and in-place variants)
- the written-to tensor doesn't share memory with any read-from tensors.

There are 4 tensors in an ternary operation:
- output (written-to)
- self (always read-from, can be written-to)
- tensor1, tensor2 (read-from)

We can reuse `binary_check_input_output_mem_overlap` and `binary_check_internal_mem_overlap` by creating binary functions and testing the following:
- output, self, tensor1
- output, self, tensor2

Something like:
```
def outplace_binop1(self, tensor1, out):
    return op(self, tensor1, torch.randn_like(tensor1), out)

def outplace_binop2(self, tensor1, out):
    return op(self, torch.randn_like(tensor2), tensor2, out)

binary_check_input_output_mem_overlap(binop1)
binary_check_input_output_mem_overlap(binop2)

// then something similar for the in-place variant
```

I think this will be more compact than writing it all like this.

Reply:
addressed